File: 410-split-array-largest-sum/410-split-array-largest-sum.cpp
class Solution {
public:
    int minimum_subarray_required(vector<int>& nums,int max_sum_allowed){
        int current_sum=0;
        int splits=0;
        for(auto ele:nums){
            current_sum+=ele;
            if (current_sum>max_sum_allowed){
                splits++;
                current_sum=ele;
                
                    
            }
        }
        return splits+1;
    }
    int splitArray(vector<int>& nums, int m) {
        int sum=0;
        int max_element=INT_MIN;
        for(auto ele:nums){
            sum+=ele;
            max_element=max(max_element,ele);
                
        }
        int left=max_element;
        int right=sum;
        while (left<right){
            int max_sum_allowed=left+(right-left)/2;
            if(minimum_subarray_required(nums,max_sum_allowed)<=m){
                right=max_sum_allowed;
            }
            else{
                left=max_sum_allowed+1;
            }
        }
        return left;
        
        
    }
};

File: 410-split-array-largest-sum/410-split-array-largest-sum.py
class Solution:
    def getminimum(self,prefix,dp,curr_index,sc,n):
        if(dp[curr_index][sc]!=-1):
            return dp[curr_index][sc]
        if sc==1:
            return prefix[n]-prefix[curr_index]
        minimum_largest_sum=float("inf")
        for i in range(curr_index,n-sc+1):
            first_split_sum=prefix[i+1]-prefix[curr_index]
            largest_split_sum=max(first_split_sum,self.getminimum(prefix,dp,i+1,sc-1,n))
            minimum_largest_sum=min(minimum_largest_sum,largest_split_sum)
            if first_split_sum>=minimum_largest_sum:
                break
        dp[curr_index][sc]=minimum_largest_sum
        return dp[curr_index][sc]
    def splitArray(self, nums: List[int], m: int) -> int:
        n=len(nums)
        dp=[[-1]*(m+1) for _ in range(n)]
        prefix=[0]*(n+1)
        for i in range(0,n):
            prefix[i+1]=prefix[i]+nums[i]
        return self.getminimum(prefix,dp,0,m,n)    